### ISSUE TITLE
[FreeBSD] Client fails when setting up pf firewall

### ISSUE THREAD

**Issue by vieira** [sshuttle/sshuttle#39]
Since aaa6062 FreeBSD (10.2-RELEASE-p7) can no longer be used as a client because the pf setup_firewall fails with `OSError: [Errno 19] Operation not supported by device` while adding an anchor rule ([methods/pf.py:222]( https://github.com/sshuttle/sshuttle/blob/1ae4fce6b3562898d8d9f4a3db29855ca7cea6e2/sshuttle/methods/pf.py#L222)). pf on FreeBSD is always behind the OpenBSD version. I am not sure if it's easier to fix this or revert to using ipfw on FreeBSD.

```
Starting sshuttle proxy.
firewall manager: Starting firewall with Python version 3.5.0
firewall manager: ready method name pf.
UDP enabled: False
Binding redirector: 12300
TCP redirector listening on ('127.0.0.1', 12300).
TCP redirector listening with <socket.socket fd=7, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('127.0.0.1', 12300)>.
Starting client with Python version 3.5.0
c : connecting to server...
c : executing: ['ssh', 'root@192.241.238.19', '--', 'P=python3.5; $P -V 2>/dev/null || P=python; exec "$P" -c \'import sys, os; verbosity=4; sys.stdin = os.fdopen(0, "rb"); exec(compile(sys.stdin.read(937), "assembler.py", "exec"))\' 2>~/sshuttle.log']
c :  > channel=0 cmd=PING len=7 (fullness=0)
c : connected.
Connected.
c : Waiting: 2 r=[7, 8] w=[8] x=[] (fullness=7/0)
c :   Ready: 2 r=[8] w=[8] x=[]
c : <  channel=0 cmd=PING len=7
c :  > channel=0 cmd=PONG len=7 (fullness=7)
c : <  channel=0 cmd=ROUTES len=34
firewall manager: Got subnets: [(2, 8, False, '216.0.0.0'), (2, 8, True, '127.0.0.0')]
firewall manager: Got nslist: []
firewall manager: Got ports: 0,12300,0,0
firewall manager: Got udp: False
firewall manager: setting up.
>> pfctl -s all

methods/pf.py@222: Before RDR anchor rule
firewall manager: undoing changes.
firewall manager: undoing IPv4 changes.
>> pfctl -a sshuttle -F all
firewall manager: undoing /etc/hosts changes.
Traceback (most recent call last):
  File "/usr/home/vieira/sshuttle/sshuttle/__main__.py", line 154, in <module>
    result = firewall.main(opt.method, opt.syslog)
  File "/usr/home/vieira/sshuttle/sshuttle/firewall.py", line 196, in main
    socket.AF_INET, subnets_v4, udp)
  File "/usr/home/vieira/sshuttle/sshuttle/methods/pf.py", line 223, in setup_firewall
    pf_add_anchor_rule(PF_RDR, "sshuttle")
  File "/usr/home/vieira/sshuttle/sshuttle/methods/pf.py", line 143, in pf_add_anchor_rule
    ioctl(pf_get_dev(), DIOCCHANGERULE, pr)
OSError: [Errno 19] Operation not supported by device
c : fatal: cleanup: ['/usr/local/bin/python3.5', '/usr/home/vieira/sshuttle/sshuttle/__main__.py', '-v', '-v', '-v', '-v', '--method', 'auto', '--firewall'] returned 1
```

**Comment by brianmay** [comment]
Apologies, I didn't know anybody was actually using this on *BSD platforms.

So wondering out loud: Is it possible to fix the pfctl method so it works on FreeBSD?

Unfortunately, I don't really understand how this code works, it was contributed. In fact, I am not even sure if it was tested on OpenBSD, however I note there are conditionals for MacOSX.

**Comment by vieira** [comment]
My knowledge of pf is somewhere between very little and non existent. With that said, I tried to make some sense of `methods/pf.py` and from what I understood there are mainly two kind of operations going on: those that use pfctl and those that write to the `/dev/pf` pseudo device using `ioctl`. The last kind seem to be causing problems.

One of the things that are done by writing to `/dev/pf` is the creation of the `anchor` and `rdr-anchor` sshuttle. That's causing the error I posted earlier. I added the anchors manually to `pf.conf` and manage to get the initial setup done, but then got another error (below) on `pf_query_nat` which also happend to be using `ioctl` to write directly to `/dev/pf`.

I agree that it would be nice to fix the pf method so it would work on OSX, FreeBSD, etc instead of having to also maintain ipfw. I will keep banging by head against the wall for a bit longer, however I am not sure something useful will come out of it. :disappointed: 

```
Starting sshuttle proxy.
firewall manager: Starting firewall with Python version 2.7.9
firewall manager: ready method name pf.
UDP enabled: False
Binding redirector: 12300
TCP redirector listening on ('127.0.0.1', 12300).
TCP redirector listening with <socket._socketobject object at 0x80431f1a0>.
Starting client with Python version 2.7.9
c : connecting to server...
c : executing: ['ssh', 'root@192.241.238.19', '--', 'P=python3.5; $P -V 2>/dev/null || P=python; exec "$P" -c \'import sys, os; verbosity=3; sys.stdin = os.fdopen(0, "rb"); exec(compile(sys.stdin.read(937), "assembler.py", "exec"))\'']
c :  > channel=0 cmd=PING len=7 (fullness=0)
server: assembling 'sshuttle' (7 bytes)
server: assembling 'sshuttle.cmdline_options' (27 bytes)
server: assembling 'sshuttle.helpers' (853 bytes)
server: assembling 'sshuttle.ssnet' (5543 bytes)
server: assembling 'sshuttle.hostwatch' (2338 bytes)
server: assembling 'sshuttle.server' (3081 bytes)
Starting server with Python version 3.5.0
 s: latency control setting = True
 s: available routes:
 s:   2/10.12.0.0/16
 s:   2/192.241.238.0/24
 s:  > channel=0 cmd=PING len=7 (fullness=0)
 s:  > channel=0 cmd=ROUTES len=34 (fullness=7)
c : Connected.
c : Waiting: 2 r=[7, 8] w=[8] x=[] (fullness=7/0)
c :   Ready: 2 r=[] w=[8] x=[]
c : mux wrote: 15/15
c : Waiting: 2 r=[7, 8] w=[] x=[] (fullness=7/0)
c :   Ready: 2 r=[8] w=[] x=[]
 s: Waiting: 1 r=[4] w=[5] x=[] (fullness=41/0)
 s:   Ready: 1 r=[] w=[5] x=[]
 s: mux wrote: 15/15
  : Waiting: 1 r=[4] w=[5] x=[] (fullness=41/0)
 s:   Ready: 1 r=[] w=[5] x=[]
 s: mux wrote: 42/42
 s: Waiting: 1 r=[4] w=[] x=[] (fullness=41/0)
c : <  channel=0 cmd=PING len=7
c :  > channel=0 cmd=PONG len=7 (fullness=7)
c : <  channel=0 cmd=ROUTES len=34
firewall manager: Got subnets: [(2, 8, False, '216.0.0.0'), (2, 8, True, '127.0.0.0')]
firewall manager: Got nslist: []
firewall manager: Got ports: 0,12300,0,0
firewall manager: Got udp: False
firewall manager: setting up.
firewall manager: setting up IPv4.
>> pfctl -s all
>> pfctl -a sshuttle -f /dev/stdin
c : mux wrote: 15/15
c : Waiting: 2 r=[7, 8] w=[] x=[] (fullness=14/0)
 s:   Ready: 1 r=[4] w=[] x=[]
 s: <  channel=0 cmd=PING len=7
 s:  > channel=0 cmd=PONG len=7 (fullness=41)
c :   Ready: 2 r=[8] w=[] x=[]
c : <  channel=0 cmd=PONG len=7
c : received PING response
c : Waiting: 2 r=[7, 8] w=[] x=[] (fullness=0/0)
 s: mux wrote: 15/15
 s: Waiting: 1 r=[4] w=[] x=[] (fullness=48/0)
 s:   Ready: 1 r=[4] w=[] x=[]
 s: <  channel=0 cmd=PONG len=7
 s: received PING response
 s: Waiting: 1 r=[4] w=[] x=[] (fullness=0/0)
c :   Ready: 2 r=[7] w=[] x=[]
c : QUERY_PF_NAT 2,6,46.101.80.172,17704,127.0.0.1,12300
 > QUERY_PF_NAT_FAILURE [Errno 19] Operation not supported by device
c : Accept TCP: 46.101.80.172:17704 -> 127.0.0.1:12300.
c : -- ignored: that's my address!
c : Waiting: 2 r=[7, 8] w=[] x=[] (fullness=0/0)
```

**Comment by brianmay** [comment]
In the last attempt by @vieira, it looks like it mostly worked, except `QUERY_PF_NAT_FAILURE` failed so we couldn't identify the original destination of the packet. I don't know if this is fixable or not.

I am not against re-adding ipfw support, although possibly it could be done a bit cleaner then before.

For example, I don't understand this `do_wait` function or this `divertsock` stuff.

If the `do_wait` is seriously required, maybe it should be a member of the `BaseMethod` class instead of being returned by `setup_firewall`.

**Comment by brianmay** [comment]
Is this fixed now #52 has been merged? Or or there still some issues remaining?

**Comment by vieira** [comment]
:+1: Working as nicely as in Linux/OSX now both as client and server. I am just unsure about what versions are supported as I have only access to 10.2.
